Update transactions doc to include bulk delete and timeouts (#240)
xataio/mdx-docs: 040-SDK/130-transaction.mdx

</TabbedCode>

#### ifVersion

The `ifVersion` parameter is used to control updates based on the version of the record. It prevents updates to a record unless the version matches the specified `ifVersion`.
We support the `ifVersion` flag to avoid overwriting unexpected versions of the record.

<TabbedCode tabs={['TypeScript', 'Python', 'Go', 'JSON']}>

```ts

> const result = await xata.transactions.run([
>   {
>     insert: {
>       table: 'titles',
>       where: { id: 'rec_cfl4g6v838og05h0iv6g', version: ifVersion },
>       record: { originalTitle: 'An updated title' }
>     }
>   }
> ]);

Reviewer: ```suggestion
const result = await xata.transactions.run([
  {
    insert: {
      table: "titles",
      record: { id: "rec_cfl4g6v838og05h0iv6g", originalTitle: "An updated title" },
      ifVersion: 0,
    },
  },
]);
```

```

```python

result = xata.records().transaction({
  "operations": [
    {
      "insert": {
        "table": "titles",
        "where": { "id": "rec_cfl4g6v838og05h0iv6g", "version": if_version },
        "record": { "originalTitle": "An updated title" }
      }
    }
  ]
})
```

```go
recordsClient, _ := xata.NewRecordsClient()
result, _ := recordsClient.Transaction(context.TODO(), xata.TransactionRequest{
  Operations: []xata.TransactionOperation{
    xata.NewInsertTransaction(xata.TransactionInsertOp{
      Table: "titles",
      Where: map[string]any{
        "id":      xata.String("rec_cfl4g6v838og05h0iv6g"),
        "version": xata.Int(ifVersion),
      },
      Record: map[string]any{
        "originalTitle": xata.String("An updated title"),
      },
    }),
  },
})
```

```json
// POST https://{workspace}.{region}.xata.sh/db/{db}:{branch}/transaction
{
  "operations": [
    {

>       "insert": {
>         "table": "titles",
>         "where": { "id": "rec_cfl4g6v838og05h0iv6g", "version": ifVersion },
>         "record": { "originalTitle": "An updated title" }
>       }

Reviewer: ```suggestion
      "insert": {
        "table": "titles",
        "record": { "id": "rec_cfl4g6v838og05h0iv6g", "originalTitle": "An updated title" },
        "ifVersion": 0
      }
```

  Operations: []xata.TransactionOperation{
    xata.NewUpdateTransaction(xata.TransactionUpdateOp{
      Table: "Users",
      Id:    "rec_cmf56ugvjdkilkpcdeu0",
      Fields: map[string]any{
        "email": xata.String("neo@matrix.com"),
        "name":  xata.String("Neo"),
      },
    }),
    xata.NewUpdateTransaction(xata.TransactionUpdateOp{
      Table: "Users",
      Id:    "rec_cmf572veoo9ahn3r96p0",
      Fields: map[string]any{
        "email": xata.String("trinity@matrix.com"),
        "name":  xata.String("Trinity"),
      },
    }),

>     // You can add additional transactions

Reviewer: ```suggestion
    // You can add additional operations
```

        "fields": { "email": "trinity@matrix.com", "name": "Trinity" }
      }
    }

>     // You can add additional transactions

Reviewer: ```suggestion
    // You can add additional operations
```

  ]
}
```

</TabbedCode>

### delete

Use `delete` to remove multiple records. Delete can operate on records from the same transaction, and will not cancel a transaction
if no record is found.

<TabbedCode tabs={['TypeScript', 'Python', 'Go', 'JSON']}>
```ts
const result = await xata.transactions.run([
  {
    delete: {
      table: 'titles',
      record: { id: 'rec_cfl4g6v00023' },
    }
  }, 
  {
    delete: {
      table: 'titles',
      record: { id: 'rec_cfl4g6v00056' },
    }
  },

>   // You can add additional transactions

Reviewer: ```suggestion
  // You can add additional operations
```

]);
```

```python
result = xata.records().transaction({
  "operations": [
    {
      "delete": {
        "table": "titles",
        "record": { "id": "rec_cfl4g6v00023" }
      }
    },
    {
      "delete": {
        "table": 'titles',
        "record": { "id": "rec_cfl4g6v00056" },
      }
    },

>     # You can add additional transactions

Reviewer: ```suggestion
    # You can add additional operations
```

  ]
})
```

```go
recordsClient, _ := xata.NewRecordsClient()
result, _ := recordsClient.Transaction(context.TODO(), xata.TransactionRequest{
  Operations: []xata.TransactionOperation{
    xata.NewDeleteTransaction(xata.TransactionDeleteOp{
      Table: "titles",
      Id:    xata.String("rec_cfl4g6v00023"),
    }),
    xata.NewDeleteTransaction(xata.TransactionDeleteOp{
      Table: "titles",
      Id:    xata.String("rec_cfl4g6v00056"),
    }),

>     // You can add additional transactions

Reviewer: ```suggestion
    // You can add additional operations
```

  }
})
```

```jsonc
// POST https://{workspace}.{region}.xata.sh/db/{db}:{branch}/transaction
{
  "operations": [
    {
      "delete": {
        "table": "titles",
        "id": "rec_cfl4g6v00023"
      }
    },
    {
      "delete": {
        "table": "titles",
        "id": "rec_cfl4g6v00056"
      }
    }

>     // You can add additional transactions

Reviewer: ```suggestion
    // You can add additional operations
```

result, _ := recordsClient.Transaction(context.TODO(), xata.TransactionRequest{
  Operations: []xata.TransactionOperation{
    xata.NewGetTransaction(xata.TransactionGetOp{
      Table:   "titles",
      Id:      xata.String("rec_cfl4g6v00023"),
      Columns: []string{"id", "originalTitle"},
    }),
    xata.NewGetTransaction(xata.TransactionGetOp{
      Table:    "titles",
      Id:       xata.String("rec_cfl4g6v00056"),
      Columns: []string{"id", "originalTitle"},
    }),
  },
})
```

```jsonc
// POST https://{workspace}.{region}.xata.sh/db/{db}:{branch}/transaction
{
  "operations": [
    {
      "get": {
        "table": "titles",
        "id": "rec_cfl4g6v00023",
        "columns": ["id", "originalTitle"]
      }
    },
    {
      "get": {
        "table": "titles",
        "id": "rec_cfl4g6v00056",
        "columns": ["id", "originalTitle"]
      }
    }
  ]
}
```

</TabbedCode>

## Limitations

- A maximum of 1000 operations are allowed in a single transaction. If you exceed this limit, you will receive an error message.
- Inserting and updating content to `file` and `file\[]` (file array) columns is not permitted with transactions.

> - To prevent timeouts with the SQL proxy during transaction deletions, delete records in small batches. Each batch should not exceed 10,000 records.

Reviewer: ```suggestion
- To prevent timeouts with the SQL proxy during delete operations with a transaction, delete records in small batches. Each batch should not exceed 10,000 records.
```

### deletes

Delete is used to remove records. Delete can operate on records from the same transaction, and will not cancel a transaction
if no record is found.

<TabbedCode tabs={['TypeScript', 'Python', 'Go', 'JSON']}>
```ts
const result = await xata.transactions.run([
  {
    delete: {
      table: 'titles',
      record: { id: 'rec_cfl4g6v00023' },
    }
  }, 
  {
    delete: {
      table: 'titles',
      record: { id: 'rec_cfl4g6v00056' },
    }
  }
  // Additional transactions can be added here
]);
```

```python
result = xata.records().transaction({
  "operations": [

>     {
>       "delete": {
>         "table": "titles",
>         "where": { "id": "rec_cfl4g6v00023" }
>       }
>     },
>     {
>     delete: {
>       table: 'titles',
>       record: { id: 'rec_cfl4g6v00056' },
>     }
>   },

Reviewer: ```suggestion
    {
      "delete": {
        "table": "titles",
        "id": "rec_cfl4g6v00023"
      }
    },
    {
      "delete": {
        "table": "titles",
        "id": "rec_cfl4g6v00056",
    }
  }
```

  ]
  # Additional transactions can be added here
})
```

```go
recordsClient, _ := xata.NewRecordsClient()
result, _ := recordsClient.Transaction(context.TODO(), xata.TransactionRequest{
  Operations: []xata.TransactionOperation{
    xata.NewDeleteTransaction(xata.TransactionDeleteOp{
      Table: "titles",
      Where: map[string]any{
        "id": xata.String("rec_cfl4g6v00023"),
      },
    }),
    xata.NewDeleteTransaction(xata.TransactionDeleteOp{
      Table: "titles",
      Where: map[string]any{
        "id": xata.String("rec_cfl4g6v00056"),
    },
  }),
   // Additional transactions can be added here
  ,
})
```

```jsonc
// POST https://{workspace}.{region}.xata.sh/db/{db}:{branch}/transaction
{
  "operations": [
    {

>       "delete": {
>         "table": "titles",
>         "where": { "id": "rec_cfl4g6v00023" }
>       }
>     },
>     {
>       "delete": {
>         "table": "titles",
>         "where": { "id": "rec_cfl4g6v00056" }
>       }

Reviewer: ```suggestion
      "delete": {
        "table": "titles",
        "id": "rec_cfl4g6v00023"
      }
    },
    {
      "delete": {
        "table": "titles",
        "id": "rec_cfl4g6v00056" 
      }
```

    }
    // Additional transactions can be added here
  ]
}
```

</TabbedCode>

### gets

Get is used to retrieve a record by ID. A get operation can retrieve records created in the same transaction but will not cancel
a transaction if no record is found.

<TabbedCode tabs={['TypeScript', 'Python', 'Go', 'JSON']}>

```ts
const result = await xata.transactions.run([
  {

>     get: {
>       table: 'titles',
>       where: { id: 'rec_cfl4g6v00023' }
>     }
>   },
>   {
>     get: {
>       table: 'titles',
>       where: { id: 'rec_cfl4g6v00056' }
>     }

Reviewer: ```suggestion
    get: {
      table: 'titles',
      id: 'rec_cfl4g6v00023',
      columns: ['id', 'originalTitle']
    }
  },
  {
    get: {
      table: 'titles',
      id: 'rec_cfl4g6v00056',
      columns: ['id', 'originalTitle']
    }
```

  }
]);
```

```python
result = xata.records().transaction({
  "operations": [

>     {
>       "get": {
>         "table": "titles",
>         "where": { "id": "rec_cfl4g6v00023" }
>       }
>     },
>     {
>       "get": {
>         "table": "titles",
>         "where": { "id": "rec_cfl4g6v00056" }
>       }
>     }

Reviewer: ```suggestion
    {
      "get": {
        "table": "titles",
        "id": "rec_cfl4g6v00023",
        "columns":["id", "originalTitle"]
      }
    },
    {
      "get": {
        "table": "titles",
        "id": "rec_cfl4g6v00056",
        "columns":["id", "originalTitle"]
      }
    }
```

  ]
})
```

```go
recordsClient, _ := xata.NewRecordsClient()
result, _ := recordsClient.Transaction(context.TODO(), xata.TransactionRequest{
  Operations: []xata.TransactionOperation{
    xata.NewGetTransaction(xata.TransactionGetOp{
      Table: "titles",
      Where: map[string]any{
        "id": xata.String("rec_cfl4g6v00023"),
      },
    }),
    xata.NewGetTransaction(xata.TransactionGetOp{
      Table: "titles",
      Where: map[string]any{
        "id": xata.String("rec_cfl4g6v00056"),
      },
    }),
  },
})
```

```jsonc
// POST https://{workspace}.{region}.xata.sh/db/{db}:{branch}/transaction
{
  "operations": [
    {
      "get": {
        "table": "titles",
        "where": { "id": "rec_cfl4g6v00023" }
      }
    },
    {
      "get": {
        "table": "titles",
        "where": { "id": "rec_cfl4g6v00056" }
      }
    }
  ]
}
```

</TabbedCode>

## Limitations

- A maximum of 1000 operations are allowed in a single transaction. If you exceed this limit, you will receive an error message.
- Inserting and updating content to `file` and `file\[]` (file array) columns is not permitted with transactions.

> - To avoid the SQL proxy timeout, delete in small batches. It is recommended to delete no more than 10,000 records per batch.

Reviewer: ```suggestion
```

I suggest we remove this from the transactions page as SQL DELETE is not a transaction and it might confuse users to find it written here
